### PR TITLE
added ranking data generation

### PR DESCRIPTION
**Pull request by AliMahmoudzadeh**
- added ranking data generation
- reformatted classification
- generate a header file (will be used in transforms)

### CODE REVIEW

File: src/common/data.py [hunk 1]
```diff
@@ -51,3 +51,105 @@ def generate(self):
...
+        # Randomly generate a well conditioned input set
+        X = self.generator.randn(self.batch_size, self.n_features)
+        y = np.dot(X, self.ground_truth) + self.bias
+
+        # Add noise
+        if self.noise > 0.0:
+            y += self.generator.normal(scale=self.noise, size=y.shape)
+
+        # create n_label_classes ranking labels
+        y = ((y - min(y))/(max(y)-min(y))*self.n_label_classes).astype(int)
```
Comment (jfomhover) [inline src/common/data.py:97]: how does that work with n_targets? it looks like n_targets is not being used in any way, but I'm curious if that should be the case here so that each class corresponds to a different regression model?

if you don't want to use that approach, then the code should not use n_targets at all?

File: src/common/data.py [hunk 2]
```diff
@@ -51,3 +51,105 @@ def generate(self):
...
+        self.n_features = n_features
+        self.n_informative = min(n_features, n_informative)
+        self.n_targets = n_targets
+        self.bias = bias
+        self.noise = noise
+        self.generator = np.random.RandomState(seed)
+
+        # Generate a ground truth model with only n_informative features being non-zeros
+        self.ground_truth = np.zeros((self.n_features, self.n_targets))
+        self.ground_truth[:n_informative, :] = 100 * self.generator.rand(self.n_informative, self.n_targets)
```
Comment (jfomhover) [inline src/common/data.py:129]: same comment here about n_targets, should we just forget about it?

File: src/common/data.py [hunk 3]
```diff
@@ -51,3 +51,105 @@ def generate(self):
...
+        Returns:
+            X (numpy.ndarray)
+            y (numpy.ndarray)
+        """
+        # Randomly generate a well conditioned input set
+        X = self.generator.randn(self.batch_size, self.n_features)
+        y = np.dot(X, self.ground_truth) + self.bias
+
+        # add query column
+        query_col = [[i // self.docs_per_query] for i in range(self.batch_size)]
```
Comment (jfomhover) [inline src/common/data.py:143]: so in this situation, each batch contains docs_per_query identical query ids. but that repeats for each batch, so ultimately the dataset contains a lot of the same query ids.

since we're not using distributed for now, maybe a solution could be to introduce a batch count and use it to generate differnt query ids for each batch?

File: src/scripts/data_processing/generate_data/generate.py [hunk 4]
```diff
@@ -93,64 +95,18 @@ def get_arg_parser(cls, parser=None):
...
-
-        X, y = make_classification(
-            n_samples=total_samples,
-            n_features=args.n_features,
-            n_informative=args.n_informative,
-            n_redundant=args.n_redundant,
-            random_state=args.random_state,
+        group_o.add_argument(
+            "--output_header",
+            required=True,
```
Comment (jfomhover) [inline src/scripts/data_processing/generate_data/generate.py:100]: could you make it required in the yaml spec but not in the script? this way, if someone runs this script locally for some reason they don't have to provide this arugment?

File: tests/scripts/test_generate_data.py [hunk 5]
```diff
@@ -126,3 +133,82 @@ def test_generate_classification_data(temporary_dir):
...
+    assert os.path.isfile(
+        os.path.join(output_inference, "inference_0.txt")
+    ), "Script generate.py should generate inference_0.txt under --output dir but did not"
+    assert not os.path.isfile(
+        os.path.join(output_inference, "inference_1.txt")
+    ), "Script generate.py should NOT generate inference_1.txt under --output dir but DID"
+
+    assert os.path.isfile(
+        os.path.join(output_header, "header.txt")
+    ), "Script generate.py should generate inference_0.txt under --output dir but did not"
```
Comment (jfomhover) [inline tests/scripts/test_generate_data.py:214]: wrong comment here, should read "header.txt" instead of inference_0.txt

File: tests/scripts/test_generate_data.py [hunk 6]
```diff
@@ -126,3 +133,82 @@ def test_generate_classification_data(temporary_dir):
     assert not os.path.isfile(
         os.path.join(output_inference, "inference_1.txt")
     ), "Script generate.py should NOT generate inference_1.txt under --output dir but DID"
+
+    assert os.path.isfile(
+        os.path.join(output_header, "header.txt")
+    ), "Script generate.py should generate inference_0.txt under --output dir but did not"
+
+def test_generate_ranking_data(temporary_dir):
```
Comment (jfomhover) [inline tests/scripts/test_generate_data.py:141]: considering your code is standardizing all 3 generation tasks (thanks!) maybe we should factorize those tests into a parametrized test. but we can leave that for another PR

File: src/scripts/data_processing/generate_data/generate.py [hunk 7]
```diff
@@ -266,10 +254,10 @@ def run(self, args, logger, metrics_logger, unknown_args):
 
         # record a metric
         logger.info(f"Generating data.")
-        if args.type == "classification":
-            self.generate_classification(args)
-        elif args.type == "regression":
-            self.generate_regression_tasks(args)
+        if args.type in ["regression", "classification", "lambdarank"]:
```
Comment (jfomhover) [inline src/scripts/data_processing/generate_data/generate.py:257]: now that all 3 types are consistent, you can even skip this test and leave it to generate_task() to figure it out, we can even remove NotImplementedError since argparse will filter out the wrong values for us (I forgot that in my initial code).

File: src/scripts/data_processing/generate_data/generate.py (hunk 4 above)
Comment (AliMahmoudzadeh) [reply src/scripts/data_processing/generate_data/generate.py:100]: done.

File: src/scripts/data_processing/generate_data/generate.py (hunk 7 above)
Comment (AliMahmoudzadeh) [reply src/scripts/data_processing/generate_data/generate.py:257]: done

File: src/common/data.py (hunk 2 above)
Comment (AliMahmoudzadeh) [reply src/common/data.py:129]: We can make the generator more flexible, but with this implementation, we are not using multiple targets. I agree with removing it now, we can add it back if we want to use it.

File: src/common/data.py (hunk 1 above)
Comment (AliMahmoudzadeh) [reply src/common/data.py:93]: I removed n_targets altogether, we can bring it back if we want to create more complicated data generation later

Comment (AliMahmoudzadeh) [reply src/common/data.py:93]: I removed all the n_targets, we can bring them back if needed in future

File: src/common/data.py (hunk 3 above)
Comment (AliMahmoudzadeh) [reply src/common/data.py]: fixed

File: tests/scripts/test_generate_data.py (hunk 5 above)
Comment (AliMahmoudzadeh) [reply tests/scripts/test_generate_data.py]: done

File: tests/scripts/test_generate_data.py (hunk 6 above)
Comment (AliMahmoudzadeh) [reply tests/scripts/test_generate_data.py:141]: I agree

**Review by jfomhover — APPROVED** [review]
